src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatMessage.cs

```diff
@@ -6,13 +6,97 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Text.Json;
 using OpenAI.Chat;
 
 namespace Microsoft.Extensions.AI;
 
 internal static partial class OpenAIModelMappers
 {
+    internal static IEnumerable<ChatMessage> FromOpenAIChatMessages(IEnumerable<OpenAI.Chat.ChatMessage> inputs, JsonSerializerOptions options)
+    {
+        // Maps all of the OpenAI types to the corresponding M.E.AI types.
+        // Unrecognized or non-processable content is ignored.
+
+        foreach (OpenAI.Chat.ChatMessage input in inputs)
+        {
+            switch (input)
+            {
+                case SystemChatMessage systemMessage:
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.System,
+                        AuthorName = systemMessage.ParticipantName,
+                        Contents = FromOpenAIChatContent(systemMessage.Content),
+                    };
+                    break;
+
+                case UserChatMessage userMessage:
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.User,
+                        AuthorName = userMessage.ParticipantName,
+                        Contents = FromOpenAIChatContent(userMessage.Content),
+                    };
+                    break;
+
+                case ToolChatMessage toolMessage:
+                    string textContent = string.Join(string.Empty, toolMessage.Content.Where(part => part.Kind is ChatMessageContentPartKind.Text).Select(part => part.Text));
```

Comment:
Why would we join up all the content parts into a single string, instead of emitting them each as separate AIContent items?

I'm not 100% certain if OpenAI's schema is documented clearly enough to say whether they would ever split a JSON value across multiple content items in the same message, or whether they may embed multiple independent function result content items in the same message. But if they ever do have multiple independent function results in the same `tool` message, it would be incorrect to concatenate their texts together.

Reply:
`FunctionResultContent` doesn't currently model splitting results into multiple parts though. What do you suggest as an alternative? Picking the first part only? Emitting separate results contents for each part? The latter doesn't feel right to me.

Reply:
Do we think we need to revisit FunctionResultContent?

Reply:
OK I think I see now, and agree with you.

Previously I thought that `ToolChatMessage` might contain multiple `Content` entries that may each represent separate function results.

However on closer inspection I see that the CallId itself only exists at the `ToolChatMessage` level, and not within any of the `Content` entries. As such it's impossible for the content entries to be separate function results, since they would have no way to indicate which call they are for.

So I agree that all the `Content` entries within a given `ToolChatMessage` must refer to the same function call result. AFAIK we likely could just return the first or even throw if there's more than one. But what you're doing with concatenation seems safe, even if it never actually has to deal with more than one content entry in practice.